R/zzz.R

```diff
@@ -35,3 +55,82 @@
   }
   packageStartupMessage(config_message)
 }
+
+python_is_compatible <- function(python_path) {
+  MIN_PYTHON_VERSION = "3.6"
+
+  Sys.setenv(RETICULATE_PYTHON = python_path)
+  python_info <- reticulate::py_discover_config()
+  Sys.unsetenv("RETICULATE_PYTHON")
+
+  if (!is.null(python_info$libpython) &&
+      utils::compareVersion(python_info$version, MIN_PYTHON_VERSION) >= 0) {
+    return(TRUE)
+  }
+  return(FALSE)
+}
+
+find_compatible_python <- function() {
+  python_path <- NULL
+  config_info <- reticulate::py_discover_config()
+  for (found_python in config_info$python_versions) {
+    if (python_is_compatible(found_python)) {
+      python_path <- found_python
+      break
+    }
+  }
+
+  if (is.null(python_path)) {
+    packageStartupMessage("No compatible version of Python (>=3.6, libpython libraries) found.")
+
+    # if on MacOS, suggest installing python3 via homebrew
+    sysname <- Sys.info()[["sysname"]]
+    if (sysname == "Darwin") {
+      suggestion_message <- paste(
+        "MacOS detected. Try installing Python3 via Homebrew (https://brew.sh/):",
+        "",
+        "  brew install python",
+        "",
+        "If Homebrew + Python are installed and problems persist, try reinstalling:",
+        "",
+        "  brew reinstall python",
+        "",
+        "After installation, make sure that Homebrew is on your $PATH.",
+        "Note that Python 3.6 or greater is required for Toolchest.",
+        sep = "\n"
+      )
+      packageStartupMessage(suggestion_message)
+    }
+
+    stop("No compatible version of Python (>=3.6, libpython libraries) found.", call. = FALSE)
+  }
+
+  return(python_path)
+}
+
+configure_virtualenv <- function(env_name, python_path) {
+  SETUPTOOLS_VERSION = "58.0.0"
+
+  # Create environment if it doesn't exist.
+  if (!reticulate::virtualenv_exists(env_name)) {
+    packageStartupMessage("Creating Python virtual environment...")
+    reticulate::virtualenv_create(env_name, python = python_path, setuptools_version = "58.0.0")
+  }
+
+  # Check the version of setuptools. Reinstall if needed.
+  version_check <- reticulate::py_run_string("reset_setuptools = False
+try:
+  from distutils.version import LooseVersion
+  import setuptools
+
+  setuptools_version = setuptools.__version__
+  reset_setuptools = (LooseVersion(setuptools_version) >= LooseVersion('58.0.2'))
+except ModuleNotFoundError:
+  pass
+")
```

Comment:
Is there a cleaner way that we could do this? E.g. a separate Python file?

Code in strings is generally undesired for maintainability.

Reply:
The only tricky part would be directing R to the path of the python file, but I think that should be doable.